### PR TITLE
Fix invalid_request_error when level3 line_items contain a negative unit cost

### PR DESCRIPTION
**Pull request by mgascam**
Fixes [721](712-gh-Automattic/woocommerce-payments-server)

#### Changes proposed in this Pull Request
Fixes `invalid_request_error` when a payment_intent is created with level3 data. If the order contains a product that has a negative price, Stripe throws an error: `Invalid non-negative integer`.

The fix proposed would be to check if there is a `line_item` in the order with a negative price when the level3 data is calculated. When that is the case, we map the data in a different way:

1. Map the calculated `unit_cost` as `discount_amount`. We have to take the absolute value since Stripe only supports positive values
2. The `unit_cost` is set to 0

Using this approach the payment_intent will be created with the level3 data. The level3 data will include all the line_items  that have a negative price, but instead of having this amount in the `unit_cost`, it will be as `discount_amount`.

#### Testing instructions

1. Go to WP Admin > Products > Add New
2. Fill the product details, setting a negative amount in the Regular Price field, and Publish the product.
3. Create an order with some products, including the product created in previous step. Note that the total amount of the order needs to be greater than 0 to be able to check out.
4. Proceed to checkout and pay the order.
5. Check that the order is processed correctly.
6. In the Stripe Dashboard, check that the created payment contains the level3 data. Please check that the `line_items` are correct. The product with a negative price should show a `unit_cost` of 0 and the `discount_amount` should match the absolute value of the product price.

**Expected level3 data in Stripe's dashboard for the payment**

Screen Shot 2022-05-10 at 12 22 02

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : [Test Instructions](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-4.2.0#fix-invalid_request_error-for-payment_intents-caused-by-level3-data-unit_cost-is-negative-value-for-discount)

### CODE REVIEW

File: includes/class-wc-payment-gateway-wcpay.php [hunk 1]
```diff
@@ -2220,6 +2220,14 @@ public function get_level3_data_from_order( WC_Order $order ): array {
 			$tax_amount      = WC_Payments_Utils::prepare_amount( $item->get_total_tax(), $currency );
 			$discount_amount = WC_Payments_Utils::prepare_amount( $subtotal - $item->get_total(), $currency );
 
+			// This is a special case, to prevent Stripe error when $unit_cost is negative. The unit cost can be
+			// negative if the item has a negative price.
+			if ( 0 > $subtotal ) {
+				// To prevent the Stripe API error, set the absolute value of the unit cost as discount.
+				$discount_amount = abs( $unit_cost );
+				$unit_cost       = 0;
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:2228]: https://stripe.com/docs/level3 says unit cost shall be greater than zero, so this approach will not work. Can you test this please before we proceed with reviews?

Comment (mgascam) [reply includes/class-wc-payment-gateway-wcpay.php:2228]: Good point. I tested it and it allowed me to create the payment_intent including the level3 data: ["pi_3KxXtsQsDOQXPzI10nBXFZqc"](https://dashboard.stripe.com/test/payments/pi_3KxXtsQsDOQXPzI10nBXFZqc). This is an excerpt of the payment_intent response from the Stripe logs:

````
 "level3": {
          "customer_reference": "190",
          "line_items": [
            {
              "discount_amount": 2000,
              "product_code": "179",
              "product_description": "You Saved 5% OFF (Discount",
              "quantity": 1,
              "tax_amount": 0,
              "unit_cost": 0
            },
            {
              "discount_amount": 0,
              "product_code": "17",
              "product_description": "Sunglasses",
              "quantity": 1,
              "tax_amount": 0,
              "unit_cost": 9000
            }
          ],
          "merchant_reference": "190",
          "shipping_address_zip": "90210",
          "shipping_amount": 100,
          "shipping_from_zip": "94110"
        },
````
It seems the documentation is not up to date or accurate?

Comment (kalessil) [reply includes/class-wc-payment-gateway-wcpay.php:2228]: THanks for testing, probably the documentation is not up to date. I'll ping stripe for updating it.

File: includes/class-wc-payment-gateway-wcpay.php [hunk 2]
```diff
@@ -2220,6 +2220,14 @@ public function get_level3_data_from_order( WC_Order $order ): array {
 			$tax_amount      = WC_Payments_Utils::prepare_amount( $item->get_total_tax(), $currency );
 			$discount_amount = WC_Payments_Utils::prepare_amount( $subtotal - $item->get_total(), $currency );
 
+			// This is a special case, to prevent Stripe error when $unit_cost is negative. The unit cost can be
+			// negative if the item has a negative price.
+			if ( 0 > $subtotal ) {
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:2225]: In this case, we should rather refactor the logic than try to patch it.
The used approach will work, but pretty quick context is lost and it'll be no clue if this block can be relocated.

To make it maintainable, the code structure shall be changed in the way to say "this is part of discount calculation, not a randomly placed condition."

In the practical field:

```php
			$description     = substr( $item->get_name(), 0, 26 );
			$quantity          = ceil( $item->get_quantity() );
			$tax_amount    = WC_Payments_Utils::prepare_amount( $item->get_total_tax(), $currency );
			if ( $subtotal >= 0 ) {
				$unit_cost             = WC_Payments_Utils::prepare_amount( $subtotal / $quantity, $currency );
				$discount_amount = WC_Payments_Utils::prepare_amount( $subtotal - $item->get_total(), $currency );
			} else {
				// It's possible to create products with negative price - represent it as free one with discount.
				$discount_amount = abs( WC_Payments_Utils::prepare_amount( $subtotal / $quantity, $currency ) );
				$unit_cost              = 0;
			}
```

**Review by kalessil — CHANGES_REQUESTED** [review]
First feedback round, let's refactor and continue with the second one.

File: includes/class-wc-payment-gateway-wcpay.php (hunk 2 above)
Comment (mgascam) [reply includes/class-wc-payment-gateway-wcpay.php:2225]: Nice improvement thanks for suggesting!

Comment (mgascam) [reply includes/class-wc-payment-gateway-wcpay.php]: Addressed in [d72732eab99cb897cf6bb4a1ffebee16c4e6b257](https://github.com/Automattic/woocommerce-payments/pull/4244/commits/d72732eab99cb897cf6bb4a1ffebee16c4e6b257)

File: tests/unit/test-class-wc-payment-gateway-wcpay.php [hunk 3]
```diff
@@ -1065,6 +1073,50 @@ public function test_full_level3_data_with_fee() {
...
+				],
+			],
+			'shipping_address_zip' => '98012',
+			'shipping_from_zip'    => '94110',
+		];
+
+		update_option( 'woocommerce_store_postcode', '94110' );
+
+		$this->mock_wcpay_account->method( 'get_account_country' )->willReturn( 'US' );
+		$mock_order   = $this->mock_level_3_order( '98012', true, true, 1, 1 );
```
Comment (kalessil) [inline tests/unit/test-class-wc-payment-gateway-wcpay.php:1114]: Can we omit the fee inclusion here: `$mock_order = $this->mock_level_3_order( '98012', false, true, 1, 1 );`?
That would make test intention clearer.

File: tests/unit/test-class-wc-payment-gateway-wcpay.php [hunk 4]
```diff
@@ -1207,14 +1259,14 @@ public function test_level3_data_bundle() {
 
 	public function test_level3_data_bundle_for_orders_with_more_than_200_items() {
 		$this->mock_wcpay_account->method( 'get_account_country' )->willReturn( 'US' );
-		$mock_order   = $this->mock_level_3_order( '98012', true, 1, 500 );
+		$mock_order   = $this->mock_level_3_order( '98012', true, true, 1, 500 );
```
Comment (kalessil) [inline tests/unit/test-class-wc-payment-gateway-wcpay.php:1262]: Can we minimize changes here: `$mock_order = $this->mock_level_3_order( '98012', true, false, 1, 500 );`?

**Review by kalessil — COMMENTED** [review]
The changes look good. Let's address the comments related to tests and go with a team-wide review.

File: tests/unit/test-class-wc-payment-gateway-wcpay.php (hunk 4 above)
Comment (mgascam) [reply tests/unit/test-class-wc-payment-gateway-wcpay.php]: Addressed in [5c3d486](https://github.com/Automattic/woocommerce-payments/pull/4244/commits/5c3d4866fba7668abdf766121e212dbc94eb87f7)

File: tests/unit/test-class-wc-payment-gateway-wcpay.php (hunk 3 above)
Comment (mgascam) [reply tests/unit/test-class-wc-payment-gateway-wcpay.php]: Addressed in [5c3d486](https://github.com/Automattic/woocommerce-payments/pull/4244/commits/5c3d4866fba7668abdf766121e212dbc94eb87f7)

File: includes/class-wc-payment-gateway-wcpay.php [hunk 5]
```diff
@@ -2214,11 +2214,17 @@ public function get_level3_data_from_order( WC_Order $order ): array {
...
-			$quantity        = ceil( $item->get_quantity() );
-			$unit_cost       = WC_Payments_Utils::prepare_amount( $subtotal / $quantity, $currency );
-			$tax_amount      = WC_Payments_Utils::prepare_amount( $item->get_total_tax(), $currency );
-			$discount_amount = WC_Payments_Utils::prepare_amount( $subtotal - $item->get_total(), $currency );
+			$description = substr( $item->get_name(), 0, 26 );
+			$quantity    = ceil( $item->get_quantity() );
+			$tax_amount  = WC_Payments_Utils::prepare_amount( $item->get_total_tax(), $currency );
+			if ( $subtotal >= 0 ) {
+				$unit_cost       = WC_Payments_Utils::prepare_amount( $subtotal / $quantity, $currency );
+				$discount_amount = WC_Payments_Utils::prepare_amount( $subtotal - $item->get_total(), $currency );
```
Comment (dmvrtx) [inline includes/class-wc-payment-gateway-wcpay.php:2222]: I wonder if it is possible for `$subtotal` to be less than `$item->get_total()`? Would it require to cover such case as a separate flow?

Comment (kalessil) [reply includes/class-wc-payment-gateway-wcpay.php:2222]: It's the order item total, not the complete order total: it's possible to create products with negative prices.

Comment (dmvrtx) [reply includes/class-wc-payment-gateway-wcpay.php:2222]: I'll rephrase my question: is it possible that `$discount_amount` calculated here will be negative? I.e.are there tests to ensure that item's subtotal is less than total? 🙂

Comment (mgascam) [reply includes/class-wc-payment-gateway-wcpay.php:2222]: Thanks @dmvrtx and @kalessil I'll look into this

Comment (mgascam) [reply includes/class-wc-payment-gateway-wcpay.php:2222]: > I'll rephrase my question: is it possible that `$discount_amount` calculated here will be negative? I.e.are there tests to ensure that item's subtotal is less than total? 🙂

Looking into `WC_Order_Item_Product` class, I found that the `set_total` method includes some logic that prevents the item's `subtotal` to be less than `total` (full code [here](https://woocommerce.github.io/code-reference/files/woocommerce-includes-class-wc-order-item-product.html#source-view.118) for reference):
````
// Subtotal cannot be less than total.
if ( '' === $this->get_subtotal() || $this->get_subtotal() < $this->get_total() ) {
	    $this->set_subtotal( $value );
}
````
If I understand that logic correctly, `$subtotal` cannot be less than `$item->get_total()` and $discount_amount will be 0 or more than 0.

**Review by kalessil — APPROVED** [review]
LGTM!

**Review by dmvrtx — APPROVED** [review]
Looks good to me!

**Pull request merged**